File: forms/ergonomic-assessment/full-stack-with-rust-axum-loco-tera-htmx-alpine/src/engine/risk_grader.rs
use super::risk_rules::all_rules;
use super::types::{AssessmentData, FiredRule, RiskLevel};
use super::utils::{calculate_composite_score, collect_likert_items};

/// Pure function: evaluates all ergonomic risk rules against assessment data.
/// Returns the risk level, composite score (0-100), and fired rules.
/// If fewer than 5 items are answered, returns "draft" status.
pub fn calculate_risk(data: &AssessmentData) -> (RiskLevel, f64, Vec<FiredRule>) {
    // Check if enough items are answered
    let items = collect_likert_items(data);
    let answered_count = items.iter().filter(|x| x.is_some()).count();

    if answered_count < 5 {
        return ("draft".to_string(), 0.0, vec![]);
    }

    // Calculate composite score (0-100, higher = more risk)
    let score = calculate_composite_score(data).unwrap_or(0.0);

    // Fire rules
    let mut fired_rules = Vec::new();
    for rule in all_rules() {
        if (rule.evaluate)(data) {
            fired_rules.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                concern_level: rule.concern_level.to_string(),
            });
        }
    }

    // Determine risk level from composite score
    let level = if score <= 25.0 {
        "low"
    } else if score <= 50.0 {
        "moderate"
    } else if score <= 75.0 {
        "high"
    } else {
        "veryHigh"
    };

    (level.to_string(), score, fired_rules)
}

File: forms/systematic-coronary-risk-evaluation-2-diabetes/front-end-form-with-svelte/src/lib/engine/risk-grader.ts
import type { AssessmentData, FiredRule, GradingResult, RiskCategory } from './types.js';
import { evaluateRules } from './risk-rules.js';
import { detectAdditionalFlags } from './flagged-issues.js';
import { isLikelyDraft } from './utils.js';

/**
 * Pure function: evaluates all risk rules against patient data.
 * Returns risk category and fired rules.
 * Risk category is the maximum among all fired rules.
 */
export function calculateRisk(data: AssessmentData): { riskCategory: RiskCategory; firedRules: FiredRule[] } {
	if (isLikelyDraft(data)) {
		return { riskCategory: 'draft', firedRules: [] };
	}

	const firedRules = evaluateRules(data);

	const riskOrder = (level: string): number => {
		switch (level) {
			case 'low':
				return 1;
			case 'medium':
				return 2;
			case 'high':
				return 3;
			default:
				return 0;
		}
	};

	let riskCategory: RiskCategory;
	if (firedRules.length === 0) {
		riskCategory = 'low';
	} else {
		const maxLevel = firedRules.reduce(
			(max, r) => (riskOrder(r.riskLevel) > riskOrder(max) ? r.riskLevel : max),
			'low'
		);
		// Map to SCORE2-Diabetes categories
		switch (maxLevel) {
			case 'high':
				riskCategory = 'veryHigh';
				break;
			case 'medium':
				riskCategory = 'high';
				break;
			default:
				riskCategory = 'moderate';
				break;
		}
	}

	return { riskCategory, firedRules };
}

/** Full grading: risk calculation + additional flags + timestamp. */
export function gradeAssessment(data: AssessmentData): GradingResult {
	const { riskCategory, firedRules } = calculateRisk(data);
	const additionalFlags = detectAdditionalFlags(data);

	return {
		riskCategory,
		firedRules,
		additionalFlags,
		timestamp: new Date().toISOString()
	};
}
